oauth2-store/src/main/java/org/xwiki/contrib/oidc/internal/OAuth2TokenStoreCache.java

```diff
@@ -0,0 +1,124 @@
+import org.xwiki.contrib.oidc.OAuth2Token;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReferenceSerializer;
+
+/**
+ * Cache for OAuth token store.
+ *
+ * @version $Id$
+ * @since 2.19.1
+ */
+@Component(roles = OAuth2TokenStoreCache.class)
+@Singleton
+public class OAuth2TokenStoreCache implements Initializable, Disposable
+{
+    @Inject
+    private CacheManager cacheManager;
+
+    private Cache<OAuth2Token> cache;
+
+    @Inject
+    private EntityReferenceSerializer<String> serializer;
+
+    @Override
+    public void initialize() throws InitializationException
+    {
+        try {
+            this.cache = this.cacheManager.createNewCache(new LRUCacheConfiguration("oidc.client.token", 1000));
+        } catch (CacheException e) {
+            throw new InitializationException("Failed to create cache with if [oidc.client.token]");
+        }
+    }
+
+    @Override
+    public void dispose() throws ComponentLifecycleException
+    {
+        this.cache.dispose();
+    }
+
+    /**
+     * Get the token from the cache if it exists.
+     *
+     * @param documentReference the document reference of the related token (where the token is stored).
+     * @param configurationName the configuration name of the related token.
+     * @return the token if it exists in the cache.
+     */
+    public Optional<OAuth2Token> get(DocumentReference documentReference, String configurationName)
+    {
+        String cacheKey = getCacheKey(documentReference, configurationName);
+        return Optional.ofNullable(cache.get(cacheKey));
+    }
+
+    /**
+     * Add a new token into the cache.
+     *
+     * @param token the token to add in the cache.
+     * @param documentReference the document reference of the related token (where the token is stored).
+     * @param configurationName the configuration name of the related token.
+     */
+    public void add(OAuth2Token token, DocumentReference documentReference, String configurationName)
+    {
+        String cacheKey = getCacheKey(documentReference, configurationName);
+        cache.set(cacheKey, token);
+    }
+
+    /**
+     * Invalidate the corresponding cache.
+     *
+     * @param documentReference the reference of the document which changed and need to be removed from the cache.
+     * @param configurationName the configuration name of the related token.
+     */
+    public void invalidateCache(DocumentReference documentReference, String configurationName)
+    {
+        String name = getCacheKey(documentReference, configurationName);
+        this.cache.remove(name);
+    }
+
+    /**
+     * Clean all entry in the token store cache.
+     */
+    public void clearCache()
+    {
+        this.cache.removeAll();
+    }
+
+    private String getCacheKey(DocumentReference documentReference, String configurationName)
+    {
+        return serializer.serialize(documentReference) + "__" + configurationName;
```

Comment:
Not sure what is the best way to build a unique cache key. @tmortagne what would you recommend ?

Reply:
Your version looks good enough to me. The important thing is to not have any collision, and I don't see any way to have a collision with the format you chose (not sure why you put two `_` by the way).

An alternative if you really don't want to have to think about the format at all would be to create some special OAuth2TokenReference which extend `DocumentReference` and adds the configuration name as reference parameter (you could imagine use that special reference for other needs too). Then you can simply use the "withparameters" `EntityReferenceSerializer`, instead of the default one. The only downside is that it's obviously a bit slower than your very simple concatenation.